selectors/selectors.md

```diff
@@ -302,67 +265,152 @@ If it's a file in some directory, you can also start at a deeper level:
 }}}}}}
 ```
 
-IPLD Schema
------------
+Schema
+------
+
+This code block describes selectors using [IPLD Schemas](../schema-layer/schemas) syntax.
 
 ```ipldsch
-# SelectorComplex is a grouping for most of the selectors, and is really just
-#  a helper for the Selector type, which introduces one more important selector
-#   to the group, but uses a different union representation to do so (for the
-#    sake of overall serial terseness).
-type SelectorComplex union {
-	| SelectAll "a"
-	| SelectFields "f" # n.b. "SelectPath" no longer exists: is a degenerate case of SelectFields.
-	| SelectIndex "i"
-	| SelectRange "r"
-	| SelectRecursive "A"
-	| SelectUnion "u"
+## SelectorEnvelope is the recommended top-level value for serialized messages
+## that don't have established existing context with marks the start of a selector:
+## it's a single-member union used to kick us towards "nominative typing".
+##
+## See https://github.com/ipld/go-ipld-prime/blob/0692e3b8cd7f231fe5d9d16a103bbbacb23dbdb5/doc/schema.md#using-schema-match-checking-as-version-detection
+## for a background on the theory behind this gentle-nominative concept.
+type SelectorEnvelope union {
+	| Selector "selector"
 } representation keyed
 
 type Selector union {
-	| SelectorComplex map
-	| SelectTrue bool
-} representation kinded
-
-# SelectTrue is a terminal selector which will not explore any more fields,
-# and marks the node it's applied to as a match.
-type SelectTrue bool
+	| Matcher "."
+	| ExploreAll "a"
+	| ExploreFields "f" # note "ExplorePath" is a degenerate case of ExploreFields.
+	| ExploreIndex "i"
+	| ExploreRange "r"
+	| ExploreRecursive "R"
+	| ExploreUnion "|"
+	| ExploreConditional "&"
+	| ExploreRecursiveEdge "@" # sentinel value; only valid in some positions.
+} representation keyed
 
-# SelectAll considers all members of the data; think of it as similar to `./*`.
-type SelectAll struct {
-	next Selector (alias ":")
+## ExploreAll is similar to a `*` -- it traverses all elements of an array,
+## or all entries in a map, and applies a next selector to the reached nodes.
+type ExploreAll struct {
+	next Selector (alias ">")
```

Comment:
Off topic, regarding schemas, I think we have to change "alias" to something else. It seems to be a translation, not an alternative:

> alias (n): an assumed or **additional** name that a person (such as a criminal) **sometimes uses**

In this case, is it possible that a block with "next" will resolve here and be what's used if data were serialised according to this schema? Or is it intended that ">" would be the most common (or only) form at the data model layer?

Reply:
Agree.  Someone suggested to me last night that in the words used by Rust's Serde system, this would be called "rename".  That would probably do a better job describing what's going on here than "alias".

---

Separately: I think it might be neat to have a tool that does transforms *on schemas*, which we could use to (for example) drop all the 'rename's.  Dropping all the 'rename's in a schema would produce a topologically/cardinalically equivalent schema -- meaning we could map data described by one schema into the transformed schema guaranteed losslessly -- and then we could do things like easily have compact representations on the wire, using one schema, and flip it into more "debug-friendly" representations at will.

(This is all something we could also build special features for, but I think it might just "fall out naturally" from the rest of the foundational work, which would be neat.)

That's an idea for "later", though.

Reply:
The idea of "exploreRecursive" is it'll actually do most of that faff for you.  The following should be a way to say "match everything":

```yaml
exploreRecursive:
  maxDepth: 999
  sequence:
    exploreUnion:
      - exploreAll:
        next:
          exploreRecursiveEdge: {}
      - matcher: {}
```

One could also skip the union and the `Matcher` and still have a selector that explores the whole graph, but doesn't actively 'match' any nodes of it... which is perhaps a subtle distinction, but think of it like a regexp that doesn't have any parens in it.

(Internally in a library implementation: there's almost certainly going to be code which uses unions to carry around the partially evaluated exploreRecursive, if I guess rightly.  But I dunno, maybe there will be other clearer ways to implement that.  And in any event, that's an internal detail.)

...ah.  Also `stopAt` should be optional.  That's a bug, I think.

Reply:
re: notation:

I was writing them as JSON for a while as a path of least resistance, but I've started to notice the amount of closing curly-braces I accumulate gets pretty wicked, to the point that I started writing it in a very lispesque way.  JSON certainly works; I'd also consider [Indentation-sensitive S-exprs](https://srfi.schemers.org/srfi-49/srfi-49.html)!

*cough* or YAML seems particularly handy and fine.  Even if we don't have an official IPLD codec for yaml, it's pretty trivial to mentally flip it into JSON and go from there, and it saves a lot of `}`'s.  (Though one moderate downer with yaml is that a terminal `Matcher` with no value for both of the optional fields serializes as `matcher: {}`, which is perhaps a bit odd on the eyes.  Eh.  C'est la vie.)

Honestly, I don't care.  As long as it's obvious enough how it maps onto the IPLD Data Model (or JSON, which is then obvious enough to carry there), it's fine by me.

We've also been sketching a lot of stuff which refers to the full names of fields (and full names of types, when it comes to unions) rather than honoring the short alias/renames in this schema.  I think that's fine for documentation and example purposes (and making schema tooling explicitly support things like "actually, ignore the renames" is something I'd like to do someday, incidentally), though we should make a footnote somewhere about doing that, since it's otherwise potentially a bit confusing.